src/lib/x509/x509_builder.cpp

```diff
@@ -0,0 +1,272 @@
+         }
+         for(const uint32_t ipv4 : m_ipv4) {
+            subject_alt.add_ipv4_address(ipv4);
+         }
+
+         return std::make_unique<Cert_Extension::Subject_Alternative_Name>(subject_alt);
+      }
+
+      X509_DN m_subject_dn;
+      Extensions m_extensions;
+      Key_Constraints m_usage;
+      std::vector<std::string> m_email;
+      std::vector<std::string> m_dns;
+      std::vector<std::string> m_uri;
+      std::vector<std::string> m_xmpp;
+      std::vector<uint32_t> m_ipv4;
+      std::vector<OID> m_ext_usage;
+      bool m_is_ca_request = false;
+      std::optional<size_t> m_path_limit;
+};
+
+CertificateParametersBuilder::CertificateParametersBuilder() :
+      m_state(std::make_unique<CertificateParametersBuilder::State>()) {}
+
+CertificateParametersBuilder::CertificateParametersBuilder(CertificateParametersBuilder&& other) noexcept = default;
+
+CertificateParametersBuilder::~CertificateParametersBuilder() = default;
+
+X509_Certificate CertificateParametersBuilder::into_self_signed_cert(std::chrono::system_clock::time_point not_before,
+                                                                     std::chrono::system_clock::time_point not_after,
+                                                                     const Private_Key& key,
+                                                                     RandomNumberGenerator& rng,
+                                                                     std::optional<std::string_view> hash_fn,
+                                                                     std::optional<std::string_view> padding) {
+   auto signer_p = X509_Object::choose_sig_format(key, rng, hash_fn.value_or("SHA-256"), padding.value_or(""));
+   auto& signer = *signer_p;
+
+   const AlgorithmIdentifier sig_algo = signer.algorithm_identifier();
+   BOTAN_ASSERT_NOMSG(sig_algo.oid().has_value());
+
+   Extensions extensions = m_state->finalize_extensions(key);
+
+   const std::vector<uint8_t> pub_key = key.subject_public_key();
+   auto skid = std::make_unique<Cert_Extension::Subject_Key_ID>(pub_key, signer.hash_function());
+
+   extensions.add_new(std::make_unique<Cert_Extension::Authority_Key_ID>(skid->get_key_id()));
+   extensions.add_new(std::move(skid));
+
+   const auto& subject_dn = m_state->subject_dn();
+
+   return X509_CA::make_cert(
+      signer, rng, sig_algo, pub_key, X509_Time(not_before), X509_Time(not_after), subject_dn, subject_dn, extensions);
+}
+
+PKCS10_Request CertificateParametersBuilder::into_pkcs10_request(const Private_Key& key,
+                                                                 RandomNumberGenerator& rng,
+                                                                 std::optional<std::string_view> challenge_password,
+                                                                 std::optional<std::string_view> hash_fn,
+                                                                 std::optional<std::string_view> padding) {
```

Comment:
I think `challenge_password` will ultimately be kept for PKCS#10, so a small nit would be that I think it should go at the end, just so that it's consistent with the python binding where I'll end up doing `(..., hash_fn, padding=None, challenge_password=None)`

Reply:
Though I suppose it would make sense for `hash_fn` to also default to None, like `into_self_signed_cert` already does essentially. I still think they should be ordered by descending important-ness.

Reply:
hash, padding, challenge does seem like the correct ordering in terms of what a user is actually going to want to specify